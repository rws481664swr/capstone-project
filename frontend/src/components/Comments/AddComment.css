.LabeledInput.comment{
    font-size: 1.2rem;
    padding:0;
    margin:0;
}

#add-comment-button.Button{
    font-size: 1rem;
}