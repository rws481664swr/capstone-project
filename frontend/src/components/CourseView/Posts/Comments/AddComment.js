import {useState} from "react";
import useAxios from "../../../../api";

const AddComment = ({add}) => {
    const [content, setContent] = useState('')
    const {post} = useAxios()
    const clear = () => setContent('')
    const submit = async (e) => {
        e.preventDefault()
        add(content)
        clear()
    }
    return <>
        <input
            type="text"
            value={content}
            onChange={e => setContent(e.target.value)}
        />
        <button onClick={submit}>Comment</button>
    </>
    //TODO
}
export default AddComment