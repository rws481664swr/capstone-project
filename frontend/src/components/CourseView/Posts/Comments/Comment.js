import {useGlobalContext} from "../../../../state/contexts/GlobalContext";

const Comment=({comment:{_id,post, username, content, timestamp},remove})=>{
    const{username:uname}=useGlobalContext()
    return <>Comment: {content} by {username} on {timestamp} {
        uname===username && <button onClick={()=>remove(_id)}>Delete</button>
    }</>
}
export default Comment
