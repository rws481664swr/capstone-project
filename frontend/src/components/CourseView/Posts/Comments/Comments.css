.Comments-delete-button {
  color: #ff0000;
  cursor: pointer;
  font-size: 1rem;
  border: none;
    background: none;
  margin-left: 0.5rem;
  text-align: center;
}
.Comments_Comment{
    margin-top: .2rem;
    margin-bottom: .2rem;
    padding: .5rem;

    /*border: 1px solid #ccc;*/
    border-left: 1px solid #ccc;
    /*border-radius: 4px;*/
    /*background-color: #f8f8f8;*/
    font-weight: 400;
    font-size: 1rem;
    width:90%;
}
.Comments-CredentialsContainer{
    margin-top:1rem;
}

.Comments_Comment_Text_container{
  display: flex;
    justify-content: space-between;
}

.centered-div{
    display: flex;
    justify-content: center;
}
.Comments_Commented-By{
    font-size: .8rem;

    font-weight: 100;
    margin-bottom: 0.5rem;

}
.between{
    padding-left:20px;
    /*display: flex;*/
    /*justify-content: center;*/
}