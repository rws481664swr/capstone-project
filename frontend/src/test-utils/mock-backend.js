import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}