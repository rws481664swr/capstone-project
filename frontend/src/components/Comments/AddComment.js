import {useState} from "react";
import useFlash from "../../hooks/form/useFlash";
import './Comments.css'
import './AddComment.css'
import LabeledInput from "../General/LabeledInput/LabeledInput";
import Button from "../General/Button/GenericButton/Button";
const AddComment = ({add}) => {
    const [content, setContent] = useState('')
    const  [toRender, flash]= useFlash()
    const submit = async (e) => {
        e.preventDefault()
        if(!content.trim()) return flash('Comment is required')
        add(content)
        setContent('')
    }
    return <>
    {toRender}
        <LabeledInput
            id={'add-comment'}
            label={'Comment'}
            value={content}
            onChange={e => setContent(e.target.value)}
            className={'LabeledInput comment'}
            name={'content'}
            />
        <Button id={'add-comment-button'} onClick={submit}>Add a Comment</Button>
    </>
    //TODO
}
export default AddComment