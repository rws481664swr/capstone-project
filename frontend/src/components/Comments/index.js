import AddComment from "./AddComment";
import Comment from './Comment'
import React, {useEffect, useReducer} from "react";
import {REMOVE} from "../../state/actions/actions";
import useAxios from "../../hooks/ajax/useAxios";
import commentsReducer from "../../state/redux/commentsReducer";
import useToggle from "../../hooks/state/useToggle";
import './Comments.css'

function useComments(post, dispatch) {


}

const useCommentsReducer = (post) => {

    const [comments, dispatch] = useReducer(commentsReducer, [])
    const axios = useAxios()

    useEffect(() => {
        (async () => {
            if (!post) return
            console.log('POSTID', post._id)
            const payload = await axios.get(`comments/${post._id}`)
            dispatch({type: "SET_STATE", payload})
        })()
    }, [post, dispatch])

    const add = async (content) => {

        const payload = await axios.post('comments', {
            post: post._id,
            content
        }, [dispatch, axios])
        dispatch({type: "ADD_COMMENT", payload})
    }
    const remove = async (id) => {
        await axios.delete(`comments/${id}`)
        dispatch({type: "REMOVE_COMMENT", id})
    }

    return [comments, add, remove]
}
const Comments = ({post}) => {
    const [commentsVisible, toggleComments] = useToggle()
    const [comments, add, remove] = useCommentsReducer(post)
    if (!post) return null
    return <>
        <div>
         <span className={'PostDisplay-ShowComments'}
               onClick={toggleComments}>
                        {commentsVisible ? 'Show' : 'Hide'}
             &nbsp;Comments
                    </span>
        </div>
        {!commentsVisible &&
            <div className={'Comments-CredentialsContainer'}>
                {comments.length > 0
                    ?
                    comments.map(c =>
                        <Comment key={c._id}
                                 comment={c}
                                 remove={() => remove(c._id)}/>)

                    : 'No Comments'}
                <AddComment add={add}/>
            </div>}
    </>
}
export default Comments
