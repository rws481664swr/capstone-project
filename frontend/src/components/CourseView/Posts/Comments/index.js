import AddComment from "./AddComment";
import Comment from './Comment'
import React, {useCallback, useEffect, useReducer} from "react";
import {REMOVE} from "../../../../state/actions/posts";
import useAxios from "../../../../api";
import commentsReducer from "../../../../state/redux/commentsReducer";
import useToggle from "../../../../hooks/useToggle";
import './Comments.css'
const useCommentsReducer = (post) => {

    const [comments, dispatch] = useReducer(commentsReducer, [])
    const axios = useAxios()

    useEffect(() => {
        (async () => {
            if (!post) return
            const payload = await axios.get(`comments`, post._id)
            console.log(payload, typeof payload, Array.isArray(payload))
            dispatch({type: "SET_STATE", payload})
        })()
    }, [post, dispatch])
    const add =useCallback (async (content) => {
        const payload = await axios.post('comments', {
            post: post._id,
            content
        },[dispatch,axios])
        dispatch({type: "ADD_COMMENT", payload})
    },[dispatch,axios])
    const remove = useCallback( async (id) => { //TODO
        await axios.delete('comments', id)
        dispatch({type: REMOVE, id})
    },[dispatch,axios])
    return [comments, add, remove]
}
const Comments = ({post}) => {
    const [commentsVisible, toggleComments] = useToggle()
    const [comments, add, remove] = useCommentsReducer(post)
    if (!post) return null
    return <>
        <div>
         <span className={'PostDisplay-ShowComments'}
             onClick={toggleComments}>
                        {commentsVisible ? 'Show' : 'Hide'}
             &nbsp;Comments
                    </span>
        </div>
        {!commentsVisible&&
            <div className={'Comments-CredentialsContainer'}>
            {comments.length > 0
                ?
                comments.map(c =>
                    <Comment key={c._id}
                             comment={c}
                             remove={() => remove(c._id)}/>)

                : 'No Comments'}
                <AddComment add={add}/>
        </div>}
    </>
}
export default Comments
