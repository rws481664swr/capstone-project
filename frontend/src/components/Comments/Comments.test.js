import {act, fireEvent, render} from '@testing-library/react';
import Comments from "./index";
import getData from '../../mock_data/data'
import {MOCK_TOKEN} from "../../../mock-backend";
import axios from "axios";
import MockContext from "../../test-utils/MockContext";

let post, comment
beforeEach(() => {
    const {postsList} = getData()
    post = postsList[0]
    comment =
        {
            "post": post._id,
            "username": "user",
            "content": "Culpa nisi delectus voluptate.",
            "timestamp": "2023-05-24T15:56:23.258Z",
            "_id": "646e33a759eab5cd39bbb8d3",
            "__v": 0
        }
})
jest.mock('axios', () => ({
    get: jest.fn(),
    post: jest.fn(),
    put: jest.fn(),
    delete: jest.fn()
}))
const axiosMock =
    async (url,body,config) => {
        if (url.includes('comments')) {
            return {data: [comment]}
        } else if (url.includes('posts')) {
            return {data: [post]}
        } else if (url.includes('auth/token')) {
            return {data: {token: MOCK_TOKEN.current}}
        }
        throw new Error()
    }
it('should render Comments without crashing', () => {
    axios.get.mockImplementation(axiosMock)
    render(
        <MockContext>
        <Comments post={post}/>
        </MockContext>)
})


it('should render Comments Snapshot', () => {
    axios.get.mockImplementation(axiosMock)
    const {asFragment} = render(
        <MockContext>
        <Comments post={post}/>
        </MockContext>)
    expect(asFragment()).toMatchSnapshot()
} )
it('should add a comment', async () => {
    axios.get.mockImplementation(axiosMock)
    axios.post.mockResolvedValue({data: 'comment'})
    const {getByText
        ,getByLabelText , getByPlaceholderText} = render(
        <MockContext>
        <Comments post={post}/>
        </MockContext>)
    await act(async () => {
        fireEvent.click(getByText("Show Comments"))

    })

await act (()=>    fireEvent.change(getByLabelText("Comment"), {target: {value: 'new comment'}}))
    const submitButton = getByText("Add a Comment")

    await act (()=> fireEvent.click(submitButton))
    expect(axios.post).toHaveBeenCalled()
})

