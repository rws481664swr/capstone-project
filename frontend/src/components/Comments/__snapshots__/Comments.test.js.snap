// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Comments Snapshot 1`] = `
<DocumentFragment>
  <div>
    <span
      class="PostDisplay-ShowComments"
    >
      ShowÂ Comments
    </span>
  </div>
</DocumentFragment>
`;
